docs/design/runtime-state.md

```diff
@@ -0,0 +1,157 @@
+
+An extra `RuntimeData` parameter in resource CRUD functions may make them harder to unit test.
+
+### Option 1b: Resource or Provider type
+
+`RuntimeData` could be included in the `Resource` or `Provider` type, so that it could be accessed from within provider `Resource` methods (assuming the resource instance contains a reference to the provider instance if data is stored there). 
+
+As detailed in the [Structs and Interfaces](./structs-interfaces.md) design document, this approach would require additional tradeoffs if resources/providers are defined as interfaces in the framework: provider developers may need to implement getters on their resource/provider struct types, or at the very least embed `framework.RuntimeData` in their structs, with the following provider code:
+
+```go
+type provider {
+  framework.RuntimeData
+}
+```
+
+## Design option 2: `provider.Client`
+
+There are two types of runtime data: provider-owned, and framework-owned. In Design Option 1 above, both are stored in the `RuntimeData` struct, which means that the provider-owned data must have type `interface{}`, since there is no way for the provider to supply a type for that data. Instead, it must assert the type of `RuntimeData.ConfigureProviderData` in a similar way to `meta` in `helper/schema` (see History above).
+
+If we want the provider to be able to define the type of provider-owned runtime data, the provider must be able to define its own struct type `Provider` (see the [Structs and Interfaces](./structs-interfaces.md) design document for why this is not possible if the provider is an instance of a `framework.Provider` struct), on which, for example, the API client could be stored, so the provider code could look like the following:
+
+```go
+type provider struct {
+  Client *github.Client
+}
+```
+
+The provider code would need to set this field in the configure func:
+
+```go
+func (p Provider) Configure(ctx context.Context, req *tfprotov5.ConfigureProviderRequest, res *tfprotov5.ConfigureProviderResponse) {
+  // use config values from req to configure a client "c" of type github.Client
+  
+  p.Client = c
+  
+  return nil
+}
+```
+
+Then, assuming that the provider instance was available to resource instances, being injected in the resource factory for example, resources could use the configured client during CRUD functions:
+
+```go
+func (r myResource) Create(ctx context.Context, req framework.ResourceCreateRequest, resp framework.ResourceCreateResponse) {
+  // determine correct parameters "params" for API request from config and state
+  
+  r.Provider.Client.CreateResource(params)
+}
+```
+
+The signatures of `Configure` and `Create` functions here are illustrative examples.
+
+This approach has the clear advantage, over Option 1, that the provider API client is now strongly typed. No type assertions are necessary, and it is clear at compile time prior to any resource implementations whether `provider.Client` is of the correct type.
```

Comment:
The only trade-off perhaps worth mentioning for this is that it may (depending on how `Provider` is surfaced to resources) require providers to use mutexes to write concurrency-safe code, or for us to do some mutex trickery ourselves (e.g., creating a RWMutex and locking it before ConfigureProvider, unlocking it after, and read-locking/unlocking it before/after every CRUD call).

Reply:
Agree, great call.